chatsky/llm/prompt.py

```diff
@@ -17,35 +19,98 @@ class PositionConfig(BaseModel):
     Lower number prompts will go before higher number prompts in the
     LLM history.
     """
-
     system_prompt: float = 0
     history: float = 1
     misc_prompt: float = 2
     call_prompt: float = 3
     last_turn: float = 4
 
 
-class Prompt(BaseModel):
+class BasePrompt(BaseModel, ABC):
     """
-    LLM Prompt.
-    Provides position config and allow validating prompts from message or string.
+    Base class for all prompts.
+    Provides common interface for getting Langchain messages.
     """
-
-    message: AnyResponse
     position: Optional[float] = None
+    @abstractmethod
+    async def to_langchain_messages(
+        self, 
+        ctx: Context, 
+        source: str = "human",
+        position_config: Optional[PositionConfig] = None
+    ) -> List[Union[HumanMessage, SystemMessage, AIMessage]]:
+        """
+        Convert this prompt to a list of Langchain messages.
+        """
+        raise NotImplementedError
+
+class Prompt(BasePrompt):
     """
-    Position for this prompt.
-    If set to None, will fallback to either :py:attr:`~PositionConfig.system_prompt`,
-    :py:attr:`~PositionConfig.misc_prompt`, :py:attr:`~PositionConfig.call_prompt`
-    depending on where the type of this prompt.
+    Zero‑shot prompt: wraps a string, Message or BaseResponse
     """
-
+    message: AnyResponse
     def __init__(self, message: Union[MessageInitTypes, BaseResponse], position: Optional[float] = None):
         super().__init__(message=message, position=position)
-
     @model_validator(mode="before")
     @classmethod
     def validate_from_message(cls, data):
         if isinstance(data, (str, Message, BaseResponse)):
             return {"message": data}
         return data
+
+    async def to_langchain_messages(
+        self,
+        ctx: Context,
+        source: str = "human",
+        position_config: Optional[PositionConfig] = None
+    ) -> List[Union[HumanMessage, SystemMessage, AIMessage]]:
+        from chatsky.llm.langchain_context import message_to_langchain
+        if isinstance(self.message, BaseResponse):
+            msg = await self.message(ctx)
+        elif isinstance(self.message, Message):
+            msg = self.message
+        else:
+            msg = Message(text=str(self.message))
```

Comment:
Since `self.message` is guaranteed to be `AnyResponse` it has the `__call__` method, which returns a message so all of this code can be much shorter.